arangod/Aql/IResearchViewExecutor.tpp

```diff
@@ -1595,79 +1769,96 @@ size_t IResearchViewExecutor<ExecutionTraits>::skip(size_t limit,
 size_t IResearchViewExecutor<ExecutionTraits>::skipAll(IResearchViewStats&) {
-  TRI_ASSERT(this->_indexReadBuffer.empty());
   TRI_ASSERT(this->_filter);
   size_t skipped = 0;
   auto const count = this->_reader->size();
-  if (_readerOffset >= count) {
+  if (_segmentOffset > count) {
     return skipped;
   }
+  auto seal = irs::Finally([&]() noexcept {
+    _segmentOffset = count + 1;
+    this->_indexReadBuffer.clear();
+  });
   if (this->infos().filterConditionIsEmpty()) {
     skipped = this->_reader->live_docs_count();
-    TRI_ASSERT(_totalPos <= skipped);
-    skipped -= std::min(skipped, _totalPos);
-    _readerOffset = count;
-    _currentSegmentPos = 0;
+    size_t totalPos = std::accumulate(
+        _segmentReaders.begin(), _segmentReaders.end(), (size_t)0,
+        [](size_t acc, auto const& r) { return acc + r.totalPos; });
+    TRI_ASSERT(totalPos <= skipped);
+    skipped -= std::min(skipped, totalPos);
   } else {
     auto const approximate = this->infos().countApproximate();
-    for (; _readerOffset != count; ++_readerOffset, _currentSegmentPos = 0) {
-      if (!_itr && !resetIterator()) {
+    // possible overfetch due to parallelisation
+    skipped = this->_indexReadBuffer.size();
+    for (auto& r : _segmentReaders) {
+      if (r.itr) {
+        skipped += calculateSkipAllCount(approximate, r.currentSegmentPos,
+                                         r.itr.get());
+        r.itr.reset();
+        r.doc = nullptr;
+        r.currentSegmentPos = 0;
+      }
+    }
+    auto& reader = _segmentReaders.front();
+    while (_segmentOffset < count) {
+      reader.readerOffset = _segmentOffset++;
+      if (!resetIterator(reader)) {
         continue;
       }
-      skipped +=
-          calculateSkipAllCount(approximate, _currentSegmentPos, _itr.get());
-      _itr.reset();
-      _doc = nullptr;
+      skipped += calculateSkipAllCount(approximate, 0, reader.itr.get());
+      reader.itr.reset();
+      reader.doc = nullptr;
+      reader.currentSegmentPos = 0;
     }
   }
   return skipped;
 }
 
-template<typename ExecutionTraits>
-void IResearchViewExecutor<ExecutionTraits>::saveCollection() {
-  // We're in the middle of a reader, save the collection in case produceRows()
-  // needs it.
-  if (_itr) {
-    // collection is constant until the next resetIterator().
-    // save it to don't have to look it up every time.
-    _segment = &this->_reader->segment(_readerOffset);
-    this->_isMaterialized = false;
-    this->_indexReadBuffer.reset();
-  }
-}
-
 template<typename ExecutionTraits>
 bool IResearchViewExecutor<ExecutionTraits>::writeRow(ReadContext& ctx,
                                                       size_t idx) {
   auto const& value = this->_indexReadBuffer.getValue(idx);
-  return Base::writeRowImpl(ctx, idx, value);
+  if constexpr (Base::isLateMaterialized) {
+    return Base::writeRowImpl(ctx, idx, value);
+  } else {
+    return Base::writeRowImpl(ctx, idx, value.value());
+  }
```

Comment:
Looks like merge artifact

Reply:
No. Why? Value is either direclty accessible or throught value() call

Reply:
Because you always can call just value

Reply:
No, that would not compile. But indeed, there is a way to simplify this part.